### ISSUE TITLE
kv: merge TxnCoordSender into internal/client.DB

### ISSUE THREAD

**Issue by tamird** [cockroachdb/cockroach#10511]
Currently, there are a bunch of hacks in TxnCoordSender with respect to contexts and cancellation. In particular, since TxnCoordSender must implement the Sender interface, it has no way of obtaining a reasonable transaction-scoped context. See https://github.com/cockroachdb/cockroach/blob/dda3ee2/pkg/kv/txn_coord_sender.go#L603:L607.

A further consequence of the above is that TxnCoordSender cannot depend on the semantic correctness of the context it is given if that context is not cancellable. See https://github.com/cockroachdb/cockroach/blob/dda3ee2/pkg/kv/txn_coord_sender.go#L729:L742.

We should hoist TxnCoordSender into the client.DB's Txn() method, which already takes a transaction-lifetime context, and remove the hacks described above in the process.

This is a pretty meaty task; good for someone looking to get acquainted with kv.

cc @andreimatei @paperstreet @bdarnell @petermattis @tschottdorf

**Comment by irfansharif** [comment]
commenting because want to follow discussions here, interesting in taking it on if nobody else is.

**Comment by andreimatei** [comment]
All yours.

If we fuse the `client.DB/Txn` and the `TxnCoordSender`, I think heartbeating of the txn record should be done until an `EndTransaction` is sent, and not based on any contexts (or in any case not based on any contexts that the called into `client.DB` has passed). We should make sure that we explicitly rollback all the transactions. SQL already does that.

Also, if you're doing this work, I think it's also worth investigating if we can get rid of the `roachpb.ExternalClient` RPC interface.

**Comment by tamird** [comment]
How can you get rid of `roachpb.ExternalClient`? That is generated due to the presence of this block in `roachpb/api.proto`:
```
service External {
  rpc Batch (BatchRequest) returns (BatchResponse) {}
}
```
Please file a separate issue if you think there's something to be done there.

**Comment by andreimatei** [comment]
I'm saying we should get rid of this RPC service, if possible. It's very related to this issue - no more remote KV clients.

**Comment by tamird** [comment]
Seriously, separate issue. Removing the remote KV clients has implications on existing CLI tools and is out of the scope of this issue.

**Comment by RaduBerinde** [comment]
Pasting some messages from a conversation a while back that may contain more context/details.

--- Radu:
>TxnCoordSender.Send expects a context that tracks the entire txn, not just the
operation we are sending. The only reason for this is to support canceling a
txn through its context: when we do NewTxn(ctx, ..), that ctx ends up being
used for the first Send and then the heartbeatLoop tracks that context's
lifetime.

>This seems hacky - one would expect Send(ctx, ..) to use a context for the
operation, not the encompassing txn. Case in point, in DBServer we can't pass
the proper context because it will lead to a txn being canceled prematurely.

>My question is - do we actually need this functionality (to cancel a txn
through its context)? Apart from a unit test for this specific thing, the
tests seem to pass without this functionality.

>If we do, perhaps the right way would be to handle cancellation in client.Txn
code (e.g. listen for ctx.Done() and send an EndTransaction if it happens).
Though we would need to spawn a goroutine per transaction just for this and I
am reluctant to add more of these.

-- Tobi:
>I agree that this is hacky, and as far as I know we don't rely on that feature for SQL any more (since SQL will send an explicit abort when the connection context closes). We also don't need this context for most of the uses of `client.Txn` (since those mostly talk to the node-local TxnCoordSender and the closure won't stop retrying without also explicitly sending an abort). That leaves the case in which we talk to a node's TxnCoordSender over RPC, which I believe is only done in a few cli commands (debug range ls, etc).
>So perhaps we could get away without this cancellation for the most part. However, conceptually it's a fact that TxnCoordSender holds on to state which is tied to the lifetime of the transaction, and what it does in the absence of a context deadline (time out after 2xHeartbeatInterval) isn't really useful. The other option (leaking memory) isn't useful either.

>If I'm looking at this from a DistSQL perspective, it also seems that TxnCoordSender gets in the way of things. Presumably you would want to talk to various TxnCoordSenders for the same transaction at the same time? Currently everything has to go through the same TxnCoordSender, due to that state.

>It seems reasonable at this point to consider making TxnCoordSender embedded in the client.Txn (which would henceforth talk directly to DistSender). This would allow for running operations in a Txn in parallel, at least to some extent, and with a context that is tied to the client lifetime.

-- Andrei:
>I don't think we need to cancel the heartbeat loop through the context any more. In fact I've recently added a TODO to remove support for this cancellation, because I agree it's hacky: https://github.com/cockroachdb/cockroach/blob/9c8e6c3ab2b6ef17870baa5d07e4636674b9e4ac/sql/session.go#L192.

>I think this support has been introduced so that we could run long transactions, and have the state in the coord sender go away if the sql connection is closed. But more recently I've made the SQL session explicitly do a rollback if that happens, which has the same effect of releasing the TxnCoordSender's state. So I think we should rely on the layers above the TxnCoordSender to always send an EndTransaction. Modulo those weird RPCs Tobi was talking about.

**Comment by tamird** [comment]
@RaduBerinde thanks for adding that, it is indeed helpful. I would caution against attempting to change the cancellation logic as part of this issue - we can do that work incrementally, and attempting to do that here will be overwhelming for a contributor unfamiliar with this code and its tests.

**Comment by a6802739** [comment]
@tamird, I want to have a see for the part of kv, could I have a try for this?

**Comment by tamird** [comment]
Sure, go ahead.
